src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs

```diff
@@ -1,87 +1,93 @@
+    //
+    // Class hierarchy:
+    //
+    // - SendReceiveTestBase                                 → file:SendReceiveTest.cs
+    //   ├─ SendReceiveTest_External
+    //   │  ├─ [*]SendReceiveTest_SharedHandler_External
+    //   │  ├─ [*]SendReceiveTest_Invoker_External
+    //   │  └─ [*]SendReceiveTest_HttpClient_External
+    //   └─ SendReceiveTest_LoopbackBase                      → file:SendReceiveTest.Loopback.cs
+    //      ├─ SendReceiveTest_Loopback
+    //      │  ├─ [*]SendReceiveTest_SharedHandler_Loopback
+    //      │  ├─ [*]SendReceiveTest_Invoker_Loopback
+    //      │  └─ [*]SendReceiveTest_HttpClient_Loopback
+    //      └─ SendReceiveTest_Http2Loopback                  → file:SendReceiveTest.Loopback.cs, SendReceiveTest.Http2.cs
+    //         ├─ [*]SendReceiveTest_Invoker_Http2Loopback
+    //         └─ [*]SendReceiveTest_HttpClient_Http2Loopback
+    // ---
+    // `[*]` - concrete runnable test classes
+    // `→ file:` - file containing the class and its concrete subclasses
+
+    public abstract class SendReceiveTestBase(ITestOutputHelper output) : ClientWebSocketTestBase(output)
     {
-        public InvokerMemorySendReceiveTest(ITestOutputHelper output) : base(output) { }
+        #region Send-receive type setup
 
-        protected override bool UseCustomInvoker => true;
-    }
-
-    public sealed class HttpClientMemorySendReceiveTest : MemorySendReceiveTest
-    {
-        public HttpClientMemorySendReceiveTest(ITestOutputHelper output) : base(output) { }
+        public static readonly object[][] EchoServersAndSendReceiveType = ToMemberData(EchoServers_Values, Enum.GetValues<SendReceiveType>());
+        public static readonly object[][] UseSslAndSendReceiveType = ToMemberData(UseSsl_Values, Enum.GetValues<SendReceiveType>());
+        public static readonly object[][] SendReceiveTypes = ToMemberData(Enum.GetValues<SendReceiveType>());
 
-        protected override bool UseHttpClient => true;
-    }
-
-    public sealed class InvokerArraySegmentSendReceiveTest : ArraySegmentSendReceiveTest
-    {
-        public InvokerArraySegmentSendReceiveTest(ITestOutputHelper output) : base(output) { }
-
-        protected override bool UseCustomInvoker => true;
-    }
+        public enum SendReceiveType
+        {
+            ArraySegment,
+            Memory
+        }
 
-    public sealed class HttpClientArraySegmentSendReceiveTest : ArraySegmentSendReceiveTest
-    {
-        public HttpClientArraySegmentSendReceiveTest(ITestOutputHelper output) : base(output) { }
+        protected Func<WebSocket, ArraySegment<byte>, WebSocketMessageType, bool, CancellationToken, Task> SendAsync { get; private set; } = null!;
+        protected Func<WebSocket, ArraySegment<byte>, CancellationToken, Task<WebSocketReceiveResult>> ReceiveAsync { get; private set; } = null!;
 
-        protected override bool UseHttpClient => true;
-    }
+        private void SelectSendReceive(SendReceiveType sendReceiveTestType)
```

Comment:
The previous approach with virtual methods seemed a bit more readable than the delegates here. If this would now lead to exponential explosion of derived classes, why not just implement the `SendAsync` as simple if - else.

I know it's all the same in the end, so if you feel the way it is makes more sense then keep it. It just feels to me like abstraction overkill (my opinion, I don't have any objective "data" to support it).